src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3410,20 +3445,7 @@ public override BoundNode VisitAddressOfOperator(BoundAddressOfOperator node)
         /// </summary>
         private bool ReportNullReferenceArgumentIfNecessary(BoundExpression argument, TypeSymbolWithAnnotations argumentType, ParameterSymbol parameter, TypeSymbolWithAnnotations paramType)
         {
-            if (argumentType.IsNullable == true)
-            {
-                if (paramType.IsReferenceType && paramType.IsNullable == false)
-                {
-                    if (!ReportNullAsNonNullableReferenceIfNecessary(argument))
-                    {
-                        ReportDiagnostic(ErrorCode.WRN_NullReferenceArgument, argument.Syntax,
-                            new FormattedSymbol(parameter, SymbolDisplayFormat.ShortFormat),
-                            new FormattedSymbol(parameter.ContainingSymbol, SymbolDisplayFormat.MinimallyQualifiedFormat));
-                    }
-                    return true;
-                }
-            }
-            return false;
+            return ReportNullableAssignmentIfNecessary(argument, paramType, argumentType, useLegacyWarnings: false, assignmentKind: AssignmentKind.Argument, target: parameter);
```

Comment:
Found the issue: the handling of arguments in object creation is different than that in calls. The former keeps conversion, while the latter strips them.
The issue is that a conversion on a argument that produces a split state is unexpected. Previously, `ReportNullReferenceArgumentIfNecessary` didn't assert an unsplit state, but re-using `ReportNullableAssignmentIfNecessary` means it now does...

Chuck, we can discuss what is the correct way of handling this in the new model for handling assignments and conversions.

Here's an example:
```C#
        [Fact]
        public void TestUnaryNegation()
        {
            var source = @"
public class C<T>
{
    C(C<object> c) => throw null;
    void M(bool b)
    {
        _ = new C<object>(!b);
    }
    public static implicit operator C<T>(T x) => throw null;
}
";
 
            var c = CreateCompilation(source, parseOptions: TestOptions.Regular8);
            c.VerifyDiagnostics( );
        }
```

```
   at System.Diagnostics.Debug.Assert(Boolean condition)
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.ReportNullableAssignmentIfNecessary(BoundExpression value, TypeSymbolWithAnnotations targetType, TypeSymbolWithAnnotations valueType, Boolean useLegacyWarnings, AssignmentKind assignmentKind, Symbol target) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 497
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.ReportNullReferenceArgumentIfNecessary(BoundExpression argument, TypeSymbolWithAnnotations argumentType, ParameterSymbol parameter, TypeSymbolWithAnnotations paramType) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 3448
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.ApplyConversion(BoundNode node, BoundExpression operandOpt, Conversion conversion, TypeSymbol targetType, TypeSymbolWithAnnotations operandType, Boolean checkConversion, Boolean fromExplicitCast, Boolean& canConvertNestedNullability) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 3031
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitConversion(BoundConversion node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2801
   at Microsoft.CodeAnalysis.CSharp.BoundConversion.Accept(BoundTreeVisitor visitor) in C:\repos\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2091
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 147
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitExpressionWithoutStackGuard(BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 304
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitExpressionWithoutStackGuard(BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 1046
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 202
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitWithStackGuard(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 296
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitAlways(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 284
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.Visit(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 266
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitRvalue(BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 611
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitArgumentEvaluate(ImmutableArray`1 arguments, ImmutableArray`1 refKindsOpt, Int32 i, Boolean preserveConditionalState) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2164
```